[docs][observability] Improvements to General Debugging guide (#51004)
ray-project/ray: doc/source/ray-observability/user-guides/debug-apps/general-debugging.rst

.. _observability-general-debugging:

> Common Cluster Issues

Author: changed the heading because this isn't really about how to debug, it's listing common pitfalls

Author: good point

This document outlines common issues encountered when running Ray in a cluster, highlighting key differences compared to running Ray locally.

Environment variables aren't passed from the Driver process to Worker processes
---------------------------------------------------------------------------------

**Issue:** When you set an environment variable at the command line that launches your Driver, it isn't propagated to all the Worker processes if the cluster has already started.

**Example:** Suppose you have a file ``baz.py`` in the directory where you run Ray, and you execute the following command:

.. literalinclude:: /ray-observability/doc_code/gotchas.py
   :language: python
   :start-after: __env_var_start__
   :end-before: __env_var_end__

**Expected behavior:** Users may expect that setting environment variables on the Driver sends them to all Worker processes as if running on a single machine, but it doesn't.

**Fix:** Enable Runtime Environments to explicitly pass environment variables. When you call ``ray.init(runtime_env=...)``, it sends the specified environment variables to the Workers.

> Alternatively, you can set the environment variable on the Driver before spawning the Workers.

Author: my understanding: if I have a running cluster, and I set an env var on the head node, that it does not propagate to the existing workers

however, at the time a new worker node is spawned, it inherits the snapshot of environment variables that are set on the head node at that moment

Author: nvm, I tested on a cluster and see that I was wrong. I've updated the docs

  @ray.remote
  def check_file():
      foo_exists = os.path.exists("/tmp/foo.txt")
      return foo_exists

  futures = []
  for _ in range(1000):
      futures.append(check_file.remote())

  print(ray.get(futures))

In this case, you might receive a mixture of True and False. If ``check_file()`` runs on the Head Node or locally, it finds the file; however, on a Worker Node, it doesn't.

**Expected behavior:** Users generally expect file references to either work consistently or to reliably fail, rather than behaving inconsistently.

**Fix:**

— Use only shared file paths for such applications. For example, a network file system or S3 storage can provide the required consistency.
— Avoid relying on local files to be consistent across machines.

> Placement Groups aren't composable

Author: I don't understand why this example works. I tried to make a minimal reproduction to improve this section, but couldn't make it work:

```python
import ray
from ray.util.placement_group import PlacementGroupSchedulingStrategy

@ray.remote(scheduling_strategy=PlacementGroupSchedulingStrategy(placement_group="a"))
def outer():
    @ray.remote(scheduling_strategy=PlacementGroupSchedulingStrategy(placement_group=None))
    def inner():
        print("Hello")
        return

    return ray.get([inner.remote() for i in range(10)])

ray.init()
ray.get(outer.remote())
```

Reviewer: ask core team?

  @ray.remote
  def check_file():
      foo_exists = os.path.exists("/tmp/foo.txt")
      return foo_exists

  futures = []
  for _ in range(1000):
      futures.append(check_file.remote())

  print(ray.get(futures))

In this case, you might receive a mixture of True and False. If ``check_file()`` runs on the Head Node or locally, it finds the file; however, on a Worker Node, it doesn't.

**Expected behavior:** Users generally expect file references to either work consistently or to reliably fail, rather than behaving inconsistently.

**Fix:**

— Use only shared file paths for such applications. For example, a network file system or S3 storage can provide the required consistency.
— Avoid relying on local files to be consistent across machines.

Placement Groups aren't composable
-----------------------------------

> **Issue:** If you call a task within a Placement Group that uses Ray Tasks, the system might fail to allocate resources properly, causing the operation to hang.

Reviewer: The phrasing is a bit weird.

how about this? "If you schedule a new task from the tasks or actors running within a Placement Group"

Author: ok done :)

> **Issue:** If you call a task within a Placement Group that uses Ray Tasks, the system might fail to allocate resources properly, causing the operation to hang.

Author: ```suggestion
**Issue:** If you schedule a new task from the tasks or actors running within a Placement Group, the system might fail to allocate resources properly, causing the operation to hang.
```
as per https://github.com/ray-project/ray/pull/51004/files/322171750facb28f33659c9d00bb39974e26c508#r1999497930

  def f():
      return 2

  print(ray.get(f.remote()))  # This should print 2.

.. testoutput::

  2

However, there are cases where modifying a remote function doesn't take effect without restarting the cluster:

> — **Imported function issue:** If ``f`` is defined in an external file (e.g., ``file.py``), and you modify its definition, re-importing the file may be ignored because Python treats the subsequent import as a no-op. A solution is to use ``from importlib import reload; reload(file)`` instead of a second import.

Author: is there a way to disable this behavior? I've had this issue when actively developing on a running cluster and it would be useful to disable this optimization.

— **Helper function dependency:** If ``f`` depends on a helper function ``h`` defined in an external file, changes to ``h`` may not propagate. The easiest solution is to restart the Ray cluster. Alternatively, you can redefine ``f`` to reload ``file.py`` before invoking ``h``:

.. testcode::

  @ray.remote
  def f():
      from importlib import reload
      reload(file)
      return file.h()

This forces the external module to reload on the Workers. Note that in Python 3, you must use ``from importlib import reload``.

Capture task and actor call sites
---------------------------------

Ray captures and displays a stack trace when you invoke a task, create an actor, or call an actor method.

To enable call site capture, set the environment variable ``RAY_record_task_actor_creation_sites=true``. When enabled:

— Ray captures a stack trace when creating tasks, actors, or invoking actor methods.

> — The captured stack trace is visible in the Ray Dashboard (under task and actor details), by running the CLI command ``ray list task --detail``, and in state API responses.

Reviewer: ```suggestion
— The captured stack trace is available in the Ray Dashboard (under task and actor details), output of the state CLI command ``ray list task --detail``, and state API responses.
```
how about this?
